File: FNV1A64_base.cs
using fnvHashFinder;
using System.Diagnostics;

internal class FNV1A64base {
    private static void Main(string[] args) {
        string? Path = string.Empty;
        string? xAsset = string.Empty;

        // You can add Direct names under any of the types
        string[] animCheck = {
"",
};
        string[] xAnim = {
"",
};
        string[] animFinalCheck = {
"",
};
        string[] imageCheck = {
"",
};
        string[] xImage = {
"",
};
        string[] materialCheck = {
"",
};
        string[] xMaterial = {
"",
};
        string[] modelCheck = {
"",
};
        string[] xModel = {
"",
};
        string[] xWpnAttachment = {
"",
};
        string[] xWpnBlueprint = {
"",
};
        string[] soundWpnCheck = {
"",
};
        string[] xWeaponSound = {
"",
};
        string[] soundOperaterCheck = {
"",
};
        string[] soundTaacomCheck = {
"",
};
        string[] xOpVox = {
"",
};
        string[] xTAACOM = {
"",
};
        string[] xGenericSound = {
"",
};
        string[] xLanguage = {
"",
};
        string[] stringCheck = {
"",
};
        string[] xString = {
"",
};

        Console.Title = "fnvHashFinder";
        Console.ForegroundColor = ConsoleColor.Yellow;
        Console.WriteLine("                         __,,,,_");
        Console.WriteLine("          _ __..-;''`--/'/ /.',-`-.");
        Console.WriteLine("      (`/' ` |  \\ \\ \\ / / / / .-'/`,_");
        Console.WriteLine("     /'`\\ \\   |  \\ | \\| // // / -.,/_,'-,");
        Console.WriteLine("    /<7' ;  \\ \\  | ; ||/ /| | \\/    |`-/,/-.,_,/')");
        Console.WriteLine("   /  _.-, `,-\\,__|  _-| / \\ \\/|_/  |    '-/.;.\'");
        Console.WriteLine("   `-`  f/ ;      / __/ \\__ `/ |__/ |");
        Console.WriteLine("    fnv `-'      |  -| =|\\_  \\  |-' |");
        Console.WriteLine("    Hash      __/   /_..-' `  ),'  //");
        Console.WriteLine("    Finder   ((__.-'((___..-'' \\__.'");
        Console.WriteLine("\nA Multi-purpose Tool that verifies pre-defined assets\n");
        Console.WriteLine("\nThis Project was made possible by:\n");
        Console.WriteLine("Scobalula, itsNatoriousB, Cortland, Amorfirion, Dark7x, DeltaDriver, pmr360, JohnWick\n");
        Console.ForegroundColor = ConsoleColor.Green;
        Console.WriteLine("[FNV1A64 BASE MODE]\n");
        Console.ForegroundColor = ConsoleColor.Yellow;
        Console.WriteLine("Specify xAsset Directory:");
        Path = Console.ReadLine();
        while (true) {
            Console.WriteLine("[1] Start Scan");
            string? search = Console.ReadLine();
            if (search == "1") {
                Console.WriteLine("\nScanning for Assets...");
                Stopwatch stopWatch = new Stopwatch();
                stopWatch.Start();
                int totalHashesFound = 0;
                Task.WaitAll(Task.Factory.StartNew(() => SearchAsset(xAsset, ref totalHashesFound)));
                stopWatch.Stop();
                TimeSpan ts = stopWatch.Elapsed;
                string elapsedTime = string.Format("{0:00}:{1:00}:{2:00}.{3:00}",
                ts.Hours, ts.Minutes, ts.Seconds, ts.Milliseconds / 10);
                Console.WriteLine("\nScan completed in " + elapsedTime);
                Console.WriteLine("Total assets found: " + totalHashesFound); }}

        void SearchAsset(string xAsset, ref int totalHashesFound) {
			int intermediateTotal = 0;
			int iterationsCompleted = 0;
			int numIterations =
                animCheck.Length * xAnim.Length * animFinalCheck.Length +
                imageCheck.Length * xImage.Length +
                materialCheck.Length * xMaterial.Length +
                modelCheck.Length * xModel.Length +
                soundWpnCheck.Length * xWeaponSound.Length * xGenericSound.Length +
                soundOperaterCheck.Length * xOpVox.Length * xLanguage.Length +
                soundTaacomCheck.Length * xTAACOM.Length * xLanguage.Length +
                stringCheck.Length * xString.Length;

            Parallel.ForEach(animCheck, animcheck => {
                Parallel.ForEach(xAnim, xanim => {
                    Parallel.ForEach(animFinalCheck, animfinalcheck => {
                        CheckAnim(animcheck + xanim + animfinalcheck, ref intermediateTotal);
                        Interlocked.Add(ref iterationsCompleted, 1); }); }); });

            Parallel.ForEach(imageCheck, imagecheck => {
                Parallel.ForEach(xImage, ximage => {
                    CheckImage(imagecheck + ximage, ref intermediateTotal);
                    Interlocked.Add(ref iterationsCompleted, 1); }); });

            Parallel.ForEach(materialCheck, materialcheck => {
                Parallel.ForEach(xMaterial, xmaterial => {
                    CheckMaterial(materialcheck + xmaterial, ref intermediateTotal);
                    Interlocked.Add(ref iterationsCompleted, 1); }); });

            Parallel.ForEach(modelCheck, modelcheck => {
                Parallel.ForEach(xModel, xmodel => {
                    CheckModel(modelcheck + xmodel, ref intermediateTotal);
                    Interlocked.Add(ref iterationsCompleted, 1); }); });

            Parallel.ForEach(soundWpnCheck, soundwpncheck => {
                Parallel.ForEach(xWeaponSound, xweaponsound => {
                    Parallel.ForEach(xGenericSound, xgenericsound => {
                        CheckSound(soundwpncheck + xweaponsound + xgenericsound, ref intermediateTotal);
                        Interlocked.Add(ref iterationsCompleted, 1); }); }); });

            Parallel.ForEach(soundOperaterCheck, soundoperatercheck => {
                Parallel.ForEach(xOpVox, xopvox => {
                    Parallel.ForEach(xLanguage, xlanguage => {
                        CheckSound(soundoperatercheck + xopvox + xlanguage, ref intermediateTotal);
                        Interlocked.Add(ref iterationsCompleted, 1); }); }); });

            Parallel.ForEach(soundTaacomCheck, soundtaacomcheck => {
                Parallel.ForEach(xTAACOM, xtaacom => {
                    Parallel.ForEach(xLanguage, xlanguage => {
                        CheckSound(soundtaacomcheck + xtaacom + xlanguage, ref intermediateTotal);
                        Interlocked.Add(ref iterationsCompleted, 1); }); }); });

            Parallel.ForEach(stringCheck, stringcheck => {
                Parallel.ForEach(xString, xstring => {
                    CheckString(stringcheck + xstring, ref intermediateTotal);
                    Interlocked.Add(ref iterationsCompleted, 1); }); });
            totalHashesFound += intermediateTotal; }

            void CheckAnim(string animName, ref int totalHashesFound) {
            string hashName = string.Format("{0:x}", Hash64baseUtil.Hash64base(animName));
            if (File.Exists(Path + "\\xanim\\" + hashName)) {
                Console.WriteLine("Found Anim: {0:x}", hashName + "," + animName);
                File.AppendAllText(Path + "\\AnimsFound.txt", hashName + "," + animName + Environment.NewLine);
                File.Delete(Path + "\\xanim\\" + hashName);
                totalHashesFound++; }}

        void CheckImage(string imageName, ref int totalHashesFound) {
            string hashName = string.Format("{0:x}", Hash64baseUtil.Hash64base(imageName));
            if (File.Exists(Path + "\\ximage\\" + hashName)) {
                Console.WriteLine("Found Image: {0:x}", hashName + "," + imageName);
                File.AppendAllText(Path + "\\ImagesFound.txt", hashName + "," + imageName + Environment.NewLine);
                File.Delete(Path + "\\ximage\\" + hashName);
                totalHashesFound++; }}

        void CheckMaterial(string materialName, ref int totalHashesFound) {
            string hashName = string.Format("{0:x}", Hash64baseUtil.Hash64base(materialName));
            if (File.Exists(Path + "\\xmaterial\\" + hashName)) {
                Console.WriteLine("Found Material: {0:x}", hashName + "," + materialName);
                File.AppendAllText(Path + "\\MaterialsFound.txt", hashName + "," + materialName + Environment.NewLine);
                File.Delete(Path + "\\xmaterial\\" + hashName);
                totalHashesFound++; }}

        void CheckModel(string modelName, ref int totalHashesFound) {
            string hashName = string.Format("{0:x}", Hash64baseUtil.Hash64base(modelName));
            if (File.Exists(Path + "\\xmodel\\" + hashName)) {
                Console.WriteLine("Found Model: {0:x}", hashName + "," + modelName);
                File.AppendAllText(Path + "\\ModelsFound.txt", hashName + "," + modelName + Environment.NewLine);
                File.Delete(Path + "\\xmodel\\" + hashName);
                totalHashesFound++; }}

        void CheckSound(string soundName, ref int totalHashesFound) {
            string hashName = string.Format("{0:x}", Hash64baseUtil.Hash64base(soundName));
            if (File.Exists(Path + "\\xsound\\" + hashName)) {
                Console.WriteLine("Found Sound: {0:x}", hashName + "," + soundName);
                File.AppendAllText(Path + "\\SoundsFound.txt", hashName + "," + soundName + Environment.NewLine);
                File.Delete(Path + "\\xsound\\" + hashName);
                totalHashesFound++; }}

        void CheckString(string stringName, ref int totalHashesFound) {
            string hashName = string.Format("{0:x}", Hash64baseUtil.Hash64base(stringName));
            if (File.Exists(Path + "\\xstring\\" + hashName)) {
                Console.WriteLine("Found String: {0:x}", hashName + "," + stringName);
                File.AppendAllText(Path + "\\StringsFound.txt", hashName + "," + stringName + Environment.NewLine);
                File.Delete(Path + "\\xstring\\" + hashName);
                totalHashesFound++; }}}}

namespace fnvHashFinder {
    class Hash64baseUtil {
        public static ulong Hash64base(string input) {
            ulong num = 14695981039346656037;
            for (int index = 0; index < input.Length; ++index)
                num = (ulong)(((long)num ^ (long)input[index]) * 1099511628211L);
            num &= 0xFFFFFFFFFFFFFFF;
            return num; }}}

File: fnv1a64base.cpp
#include "fnv1a64.h"

string path;

// You can add Direct names under any of the types
string animCheck[] = {
"",
};
string xAnim[] = {
"",
};
string animFinalCheck[] = {
"",
};
string imageCheck[] = {
"",
};
string xImage[] = {
"",
};
string materialCheck[] = {
"",
};
string xMaterial[] = {
"",
};
string modelCheck[] = {
"",
};
string xModel[] = {
"",
};
string xWpnAttachment[] = {
"",
};
string xWpnBlueprint[] = {
"",
};
string soundWpnCheck[] = {
"",
};
string xWeaponSound[] = {
"",
};
string soundOperaterCheck[] = {
"",
};
string soundTaacomCheck[] = {
"",
};
string xOpVox[] = {
// Operator Voicelines
"",
};
string xTAACOM[] = {
// TAACOM Voicelines
"",
};
string xGenericSound[] = {
"",
};
string xLanguage[] = {
"",
};
string stringCheck[] = {
"",
};
string xString[] = {
"",
};

void CheckAsset(string assetType, string assetName, int& totalHashesFound, string& foundAssetType) {
    const char* assetNameCStr = assetName.c_str();
    uint64_t assetNameHash = Hash64Util::Hash64(assetNameCStr, assetName.length());
    ostringstream ss;
    ss << hex << assetNameHash;
    string hashName = ss.str();
    if (ifstream(path + "\\" + assetType + "\\" + hashName).good()) {
        cout << "Found " << assetType << ": " << hashName << "," << assetName << "\n";
        string fileName;
        if (assetType == "xanim") {
            fileName = "AnimsFound.txt";
        }
        else if (assetType == "ximage") {
            fileName = "ImagesFound.txt";
        }
        else if (assetType == "xmaterial") {
            fileName = "MaterialsFound.txt";
        }
        else if (assetType == "xmodel") {
            fileName = "ModelsFound.txt";
        }
        else if (assetType == "xsound") {
            fileName = "SoundsFound.txt";
        }
        else if (assetType == "xstring") {
            fileName = "StringsFound.txt";
        }
        ofstream outfile(path + "\\" + fileName, ios_base::app);
        outfile << hashName << "," << assetName << "\n";
        outfile.close();
        remove((path + "\\" + assetType + "\\" + hashName).c_str());
        totalHashesFound++;
    }
}

void SearchAsset(string xAsset, int& totalHashesFound) {
    string assetType;
    int intermediateTotal = 0;
    int iterationsCompleted = 0;

    const int animIterations = sizeof(animCheck) / sizeof(animCheck[0]) * sizeof(xAnim) / sizeof(xAnim[0]) * sizeof(animFinalCheck) / sizeof(animFinalCheck[0]);
    const int imageIterations = sizeof(imageCheck) / sizeof(imageCheck[0]) * sizeof(xImage) / sizeof(xImage[0]);
    const int materialIterations = sizeof(materialCheck) / sizeof(materialCheck[0]) * sizeof(xMaterial) / sizeof(xMaterial[0]);
    const int modelIterations = sizeof(modelCheck) / sizeof(modelCheck[0]) * sizeof(xModel) / sizeof(xModel[0]);
    const int soundWpnIterations = sizeof(soundWpnCheck) / sizeof(soundWpnCheck[0]) * sizeof(xWeaponSound) / sizeof(xWeaponSound[0]) * sizeof(xGenericSound) / sizeof(xGenericSound[0]);
    const int soundOperaterIterations = sizeof(soundOperaterCheck) / sizeof(soundOperaterCheck[0]) * sizeof(xOpVox) / sizeof(xOpVox[0]) * sizeof(xLanguage) / sizeof(xLanguage[0]);
    const int soundTaacomIterations = sizeof(soundTaacomCheck) / sizeof(soundTaacomCheck[0]) * sizeof(xTAACOM) / sizeof(xTAACOM[0]) * sizeof(xLanguage) / sizeof(xLanguage[0]);
    const int stringIterations = sizeof(stringCheck) / sizeof(stringCheck[0]) * sizeof(xString) / sizeof(xString[0]);

    for (const auto& animcheck : animCheck) {
        for (const auto& xanim : xAnim) {
            for (const auto& animfinalcheck : animFinalCheck) {
                CheckAsset("xanim", animcheck + xanim + animfinalcheck, intermediateTotal, assetType);
                iterationsCompleted++;
            }
        }
    }

    for (const auto& imagecheck : imageCheck) {
        for (const auto& ximage : xImage) {
            CheckAsset("ximage", imagecheck + ximage, intermediateTotal, assetType);
            iterationsCompleted++;
        }
    }

    for (const auto& materialcheck : materialCheck) {
        for (const auto& xmaterial : xMaterial) {
            CheckAsset("xmaterial", materialcheck + xmaterial, intermediateTotal, assetType);
            iterationsCompleted++;
        }
    }

    for (const auto& modelcheck : modelCheck) {
        for (const auto& xmodel : xModel) {
            CheckAsset("xmodel", modelcheck + xmodel, intermediateTotal, assetType);
            iterationsCompleted++;
        }
    }

    for (const auto& soundwpncheck : soundWpnCheck) {
        for (const auto& xweaponsound : xWeaponSound) {
            for (const auto& xgenericsound : xGenericSound) {
                CheckAsset("xsound", soundwpncheck + xweaponsound + xgenericsound, intermediateTotal, assetType);
                iterationsCompleted++;
            }
        }
    }

    for (const auto& soundoperatercheck : soundOperaterCheck) {
        for (const auto& xopvox : xOpVox) {
            for (const auto& xlanguage : xLanguage) {
                CheckAsset("xsound", soundoperatercheck + xopvox + xlanguage, intermediateTotal, assetType);
                iterationsCompleted++;
            }
        }
    }

    for (const auto& soundtaacomcheck : soundTaacomCheck) {
        for (const auto& xtaacom : xTAACOM) {
            for (const auto& xlanguage : xLanguage) {
                CheckAsset("xsound", soundtaacomcheck + xtaacom + xlanguage, intermediateTotal, assetType);
                iterationsCompleted++;
            }
        }
    }

    for (const auto& stringcheck : stringCheck) {
        for (const auto& xstring : xString) {
            CheckAsset("xstring", stringcheck + xstring, intermediateTotal, assetType);
            iterationsCompleted++;
        }
    }
    totalHashesFound = intermediateTotal;
}

int main(const vector<wstring>& args) {
    SetConsoleTitle(L"fnvHashFinder");
    HANDLE hConsole = GetStdHandle(STD_OUTPUT_HANDLE);
    SetConsoleTextAttribute(hConsole, FOREGROUND_GREEN | FOREGROUND_RED);
    cout << "                         __,,,,_\n";
    cout << "          _ __..-;''`--/'/ /.',-`-.\n";
    cout << "      (`/' ` |  \\ \\ \\ / / / / .-'/`,_\n";
    cout << "     /'`\\ \\   |  \\ | \\| // // / -.,/_,'-,\n";
    cout << "    /<7' ;  \\ \\  | ; ||/ /| | \\/    |`-/,/-.,_,/')\n";
    cout << "   /  _.-, `,-\\,__|  _-| / \\ \\/|_/  |    '-/.;.\\'\n";
    cout << "   `-`  f/ ;      / __/ \\__ `/ |__/ |\n";
    cout << "    fnv `-'      |  -| =|\\_  \\  |-' |\n";
    cout << "    Hash      __/   /_..-' `  ),'  //\n";
    cout << "    Finder   ((__.-'((___..-'' \\__.'\n";
    cout << "\nA Versatile Tool that verifies hashes\n";
    cout << "\nThis Project was made possible by:\n";
    cout << "Scobalula, Cortland, ItsNatoriousB, Amorfirion\n";
    SetConsoleTextAttribute(hConsole, FOREGROUND_GREEN);
    cout << "\nFNV1A64 MODE\n";
    SetConsoleTextAttribute(hConsole, FOREGROUND_GREEN | FOREGROUND_RED);
    cout << "\nSpecify root directory: ";
    getline(cin, path); {
        cout << "\nScanning for Assets...\n";
        time_point start = steady_clock::now();
        int totalHashesFound = 0;
        thread t(SearchAsset, path.c_str(), ref(totalHashesFound));
        t.join();
        steady_clock::time_point end = steady_clock::now();
        duration<double> elapsed_seconds = end - start;
        int hours = static_cast<int>(elapsed_seconds.count() / 3600);
        int minutes = static_cast<int>((elapsed_seconds.count() - (hours * 3600)) / 60);
        int seconds = static_cast<int>(elapsed_seconds.count() - (hours * 3600) - (minutes * 60));
        cout << "\nScan completed in: " << setfill('0') << setw(2) << hours << ":" << setw(2) << minutes << ":" << setw(2) << seconds;
        cout << "\nTotal assets found: " << totalHashesFound << '\n';
        system("pause");
        return 0;
    }
}